core/src/replay_stage.rs

```diff
@@ -1352,6 +1371,107 @@ impl ReplayStage {
         })
     }
 
+    /// Enables alpenglow
+    /// - Clears any in progress leader blocks
+    /// - Clears any TowerBFT blocks past the genesis block
+    /// - Shutdown poh
+    /// - Start block creation loop
+    ///
+    /// Should only be called if we have received a genesis certificate on our view of the genesis block.
+    /// A false return value means that something has gone seriously wrong.
+    #[allow(clippy::too_many_arguments)]
+    fn enable_alpenglow(
+        exit: &AtomicBool,
+        my_pubkey: &Pubkey,
+        migration_status: &MigrationStatus,
+        bank_forks: &RwLock<BankForks>,
+        blockstore: &Blockstore,
+        poh_controller: &mut PohController,
+        shared_poh_bank: &SharedWorkingBank,
+        leader_schedule_cache: &LeaderScheduleCache,
+        ancestors: &mut HashMap<Slot, HashSet<Slot>>,
+        descendants: &mut HashMap<Slot, HashSet<Slot>>,
+        progress: &mut ProgressMap,
+    ) -> bool {
+        let root_bank = bank_forks.read().unwrap().root_bank();
+        warn!("{my_pubkey}: Alpenglow genesis vote has succeeded enabling alpenglow");
+
+        let genesis_block @ (genesis_slot, block_id) = migration_status.genesis_block();
+        warn!("{my_pubkey}: Alpenglow genesis block {genesis_block:?}");
+
+        let Some(genesis_bank) = bank_forks.read().unwrap().get(genesis_slot) else {
+            error!(
+                "{my_pubkey}: Attempting to enable alpenglow before receiving the genesis block"
+            );
+            return false;
+        };
+
+        if genesis_bank.block_id() != Some(block_id) {
+            error!(
+                "{my_pubkey}: Attempting to enable alpenglow but we have the wrong version of the \
+                 genesis block our version: ({genesis_slot}, {:?}), certified version \
+                 ({genesis_slot}, {block_id})",
+                genesis_bank.block_id()
+            );
+            return false;
+        }
+
+        // Reset poh to the genesis block. This has to be done first before we can clear banks to
+        // avoid any inflight issues with transaction recording.
+        Self::reset_poh_recorder(
+            my_pubkey,
+            blockstore,
+            genesis_bank,
+            poh_controller,
+            leader_schedule_cache,
+        );
+        while poh_controller.has_pending_message() || shared_poh_bank.load().is_some() {
+            std::hint::spin_loop();
+        }
+
+        // Purge all slots greater than the genesis slot from AccountsDB & blockstore
+        let slots_to_purge: Vec<Slot> = bank_forks
+            .read()
+            .unwrap()
+            .banks()
+            .iter()
+            .filter_map(|(slot, _)| (*slot > genesis_slot).then_some(*slot))
+            .collect();
+        for slot in slots_to_purge.into_iter() {
+            warn!("{my_pubkey}: Purging poh block in slot {slot}");
+            Self::purge_unconfirmed_duplicate_slot(
```

Comment:
I wonder if we should rename `fn purge_unconfirmed_duplicate_slot` now that we're leveraging it for a very different purpose